network/kademlia_test.go

```diff
@@ -424,83 +408,59 @@ func TestSuggestPeersFindPeers(t *testing.T) {
 	tk.checkSuggestPeer("01100000", 0, false)
 	tk.On("01100000")
 
+	tk.Register("11100000")
+	tk.Register("01100000")
+
 	//Bin1 should be saturated now (size == expectedSizeBin1)
-	//Next suggestions should all be bin0 peers
+	//Next suggestion should  be bin0 peers
 	tk.checkSuggestPeer("11100000", 0, false)
 	tk.On("11100000")
-	tk.checkSuggestPeer("11000000", 0, false)
-	tk.On("11000000")
 
-	//Bin0 should also  be saturated now (size == expectedSizeBin0)
+	//Bin0 should also  be saturated now (size == expectedBinSize)
 	//All bins saturated, shouldn't suggest more peers
-	tk.checkSuggestPeer("<nil>", 0, false)
-
-	//We check for change of saturationDepth
-	tk.Off("11000000")
-	tk.checkSuggestPeer("11000000", 0, true)
-	tk.On("11000000")
-
-	//We check again, this time unsaturating bin1
-	tk.Off("01110000")
-	//We don't connect to force  peer not callable
-	tk.checkSuggestPeer("01110000", 1, true)
-	tk.checkSuggestPeer("01000000", 0, false)
-	//Now no peers are callable
-	tk.checkSuggestPeer("<nil>", 0, false)
-
-	tk.On("01110000")
-	tk.Off("01110000")
-	//Peer should be callable again
-	tk.checkSuggestPeer("01110000", 1, true)
-	tk.On("01110000")
-
-	//Same with the other peer
-	tk.On("01000000")
-	tk.Off("01000000")
-	//Peer is callable again, but all bins  saturated. No peer suggested
+	tk.Register("11000000")
 	tk.checkSuggestPeer("<nil>", 0, false)
 
 	//Depth is 2
 	//Since depth is 2, bins >= 2 aren't considered saturated if peers left to connect
 	//We add addresses that fall in bin2 and bin3
-	tk.Register("00111000", "00110000")
+	tk.Register("00111000")
 	tk.Register("00011100")
 
-	//Again, should fill from shallow to deep
-	tk.checkSuggestPeer("00110000", 0, false)
-	tk.On("00110000")
+	//Again, should fill bins from shallow to deep
```

Comment:
Sorry, didn't fully understand what you meant.
If I understand correctly: when in the neighborhood, if it doesn't try to first fill with the closest peer (higher PO), then there is something wrong in the suggestPeer implementation? Is this behavior specified in the network paper? can't seem to find it

Reply:
I'm not sure if it's in the network paper. I would have to review the paper to see whether the steps in general are outlined in such detail.

My informal understanding is  that any known but unconnected neighbors should always have highest connection priority (also over the other bins), and always the closest one first. This is also what I've written in the spec SWIP for the hive component: https://github.com/ethersphere/SWIPs/pull/25 https://github.com/nolash/SWIPs/blob/hive/SWIPs/swip-hive.md

Reply:
> My informal understanding is that any known but unconnected neighbors should always have highest connection priority (also over the other bins),

This is done right now

> and always the closest one first. 

As far as I can tell, this is not being done right now (as the test shows). I believe this should be implemented in a separated PR.  For this PR I could just remove the comment to avoid saying something untrue

Reply:
That's fine, and please make an issue for this. I am sure @zelig can clear up things (and if I am wrong make the appropriate comments in the spec swip)

Reply:
Ok! will do